ompi/mca/coll/solo/coll_solo_module.c

```diff
@@ -0,0 +1,320 @@
+    }
+
+    solo_module = OBJ_NEW(mca_coll_solo_module_t);
+    if (NULL == solo_module) {
+        return NULL;
+    }
+
+    /* All is good -- return a module */
+    solo_module->super.coll_module_enable = mca_coll_solo_module_enable;
+    solo_module->super.ft_event = NULL;
+    solo_module->super.coll_allgather = NULL;
+    solo_module->super.coll_allgatherv = NULL;
+    solo_module->super.coll_allreduce = mca_coll_solo_allreduce_intra;
+    solo_module->super.coll_alltoall = NULL;
+    solo_module->super.coll_alltoallv = NULL;
+    solo_module->super.coll_alltoallw = NULL;
+    solo_module->super.coll_barrier = mac_coll_solo_barrier_intra;
+    solo_module->super.coll_bcast = mca_coll_solo_bcast_intra;
+    solo_module->super.coll_exscan = NULL;
+    solo_module->super.coll_gather = NULL;
+    solo_module->super.coll_gatherv = NULL;
+    solo_module->super.coll_reduce = mca_coll_solo_reduce_intra;
+    solo_module->super.coll_reduce_scatter = NULL;
+    solo_module->super.coll_scan = NULL;
+    solo_module->super.coll_scatter = NULL;
+    solo_module->super.coll_scatterv = NULL;
+
+    opal_output_verbose(10, ompi_coll_base_framework.framework_output,
+                        "coll:solo:comm_query (%d/%s): pick me! pick me!",
+                        comm->c_contextid, comm->c_name);
+    return &(solo_module->super);
+}
+
+/* Init the solo module on the communicator */
+static int mca_coll_solo_module_enable(mca_coll_base_module_t * module,
+                                         struct ompi_communicator_t *comm)
+{
+    /* prepare the placeholder for the array of request for invoking base module */
+    module->base_data = OBJ_NEW(mca_coll_base_comm_t);
+    if (NULL == module->base_data) {
+        return OMPI_ERROR;
+    }
+    return OMPI_SUCCESS;
+}
+
+/* Enable the solo module on the communicator lazily */
+int mca_coll_solo_lazy_enable(mca_coll_base_module_t * module, struct ompi_communicator_t *comm)
+{
+    mca_coll_solo_module_t *solo_module = (mca_coll_solo_module_t *) module;
+
+    /**
+     * Temporarily use tuned module to prevent the collective operations in this module are invoked
+     * before the initialization. 
+     */
+    int var_id;
+    int tmp_priority = 100;
+    const int *origin_priority = NULL;
+    int tmp_origin = 0;
+    mca_base_var_find_by_name("coll_tuned_priority", &var_id);
+    mca_base_var_get_value(var_id, &origin_priority, NULL, NULL);
+    tmp_origin = *origin_priority;
+    mca_base_var_set_flag(var_id, MCA_BASE_VAR_FLAG_SETTABLE, true);
+    mca_base_var_set_value(var_id, &tmp_priority, sizeof(int), MCA_BASE_VAR_SOURCE_SET, NULL);
+    comm->c_coll->coll_allreduce = ompi_coll_base_allreduce_intra_recursivedoubling;
+
+    /* Create the mpool */
+    if (mca_coll_solo_component.solo_mpool == NULL) {
+        mca_coll_solo_component.solo_mpool = OBJ_NEW(mca_coll_solo_mpool_t);
+    }
+
+    /* Create the dynamic_win */
+    ompi_win_create_dynamic((opal_info_t *) (&ompi_mpi_info_null), comm,
```

Comment:
Why do you need 2 windows when one could have been enough ? You can view the static window as the first "special" allocation on the dynamic one, and then reduce memory requirement of your module significantly.